docs/core/knowledge-bases.rst

```diff
@@ -0,0 +1,383 @@
+
+.. note::
+   There is a tutorial `here <https://blog.rasa.com/integrating-rasa-with-knowledge-bases/>`_ about how to use
+   knowledge bases in custom actions. The tutorial will explain in detail the implementation behind
+   ``ActionQueryKnowledgeBase``.
+
+
+Create a KnowledgeBase
+----------------------
+
+The data you will use to answer the user's request comes from a knowledge base.
+A knowledge base can be used to store complex data structures.
+If you just have some data points that fit into memory, you can use our ``InMemoryKnowledgeBase`` implementation.
+To initialize an ``InMemoryKnowledgeBase`` you need to provide the data in a json file.
+
+Let's take a look at an example:
+
+.. code-block:: json
+
+    {
+        "restaurant": [
+            {
+                "id": 0,
+                "name": "Donath",
+                "cuisine": "Italian",
+                "outside-seating": true,
+                "price-range": "mid-range"
+            },
+            {
+                "id": 1,
+                "name": "Berlin Burrito Company",
+                "cuisine": "Mexican",
+                "outside-seating": false,
+                "price-range": "cheap"
+            },
+            {
+                "id": 2,
+                "name": "I due forni",
+                "cuisine": "Italian",
+                "outside-seating": true,
+                "price-range": "mid-range"
+            }
+        ],
+        "hotel": [
+            {
+                "id": 0,
+                "name": "Hilton",
+                "price-range": "expensive",
+                "breakfast-included": true,
+                "city": "Berlin",
+                "free-wifi": true,
+                "star-rating": 5,
+                "swimming-pool": true
+            },
+            {
+                "id": 1,
+                "name": "Hilton",
+                "price-range": "expensive",
+                "breakfast-included": true,
+                "city": "Frankfurt am Main",
+                "free-wifi": true,
+                "star-rating": 4,
+                "swimming-pool": false
+            },
+            {
+                "id": 2,
+                "name": "B&B",
+                "price-range": "mid-range",
+                "breakfast-included": false,
+                "city": "Berlin",
+                "free-wifi": false,
+                "star-rating": 1,
+                "swimming-pool": false
+            },
+        ]
+    }
+
+The above json file contains data about restaurants and hotels.
+The json structure should contain a key for every object type, i.e. "restaurant" and "hotel".
+Every object type maps to a list of objects.
+
+Once the data are defined in a json file, called, for example, ``data.json``, you can create your
+``InMemoryKnowledgeBase``:
+
+.. code-block:: python
+
+    knowledge_base = InMemoryKnowledgeBase.load("data.json")
+
+Customize your InMemoryKnowledgeBase
```

Comment:
I feel like this section contains a lot of advanced details that most people don't need, correct? If I were looking to get started with this feature I would find all of this intimidating

Reply:
depends. if the user is using the default values, e.g. every object has an id and a name field, he does not need to do anything. if not then he should overwrite those methods.

Reply:
ok. does it work to then mention that every item in the KB should have a `name` and `id` field, and if not, they should read about customizing below?